### ISSUE TITLE
Format for {_.file}

### ISSUE THREAD

**Issue by wa0uwh** [thomasa88/ParametricText#10]
Often filenames can be very long, a string format to select a substring from the filename would be very useful.

Maybe this capability exists, but I do not know how to use it, if so please provide examples in Quick Reference.

For my usage, as in "FileName_P01", selecting the last few (ie,3) characters would be emboss on all machined parts.

FYI: my current usage on two lines are:   {_.file} v{_.version:02} and {_.date:%Y%m%d%H%M%S}

I currently use "ParametricText" on all projects.

**Comment by thomasa88** [comment]
Hi, I looked through the Python reference and the formatting only provides means to cut away the end of a string.
I'm trying to not stray too much from the standard Python formatting, but digging through some websites I think the following syntax could be seen as being in line, with the square brackets selecting a substring:
`{_.file[0:10]}` - first 10 characters (0-indexed, inclusive start, exclusive end)
`{_.file[4]}` - the fifth character (once again, 0-indexed. Maybe that is very counter-intuitive)
`{_.file[-3:]}` - last 3 characters
`{_.file[1:-1]}` - second to second last character

Do you think that would meet your needs?

An alternative would be to allow full-fledged Python scripting (`_.file.split('_')[1]`), but that would open the door to virus behavior, just like Word macros.

I'm pondering whether I should create a more non-programmer friendly alternative syntax.

**Comment by wa0uwh** [comment]
Thomas

Your suggested substring strategy will be very useful, and would meet my
needs,

Yes, I could see how Python scripting would also be very useful, but I do
not know Python well enough to understand the ramifications.

Thanks for your efforts.

Regards,
Eldon Brown

**Comment by thomasa88** [comment]
@wa0uwh Please try v2.1.0-beta.1: https://github.com/thomasa88/ParametricText/releases/tag/v2.1.0-beta.1

The following format string should give you what you want: `{_.file[-3:]}` ->  `P01`.
There are some more examples provided here: https://github.com/thomasa88/ParametricText/tree/v2.1#parameter-substrings

If you have installed the add-in from the Autodesk store, please uninstall that version first.

Also, I'm glad to hear that you find my add-in useful :).

**Comment by wa0uwh** [comment]
Thomas,

I installed v2.1.0-beta.1 from GitHup (after removing the previous), which
seems to work fine for all but one of my use cases.

I often "Save As" the Fusion file when creating a new but similar design,
which I change the last few characters of the file name as it is a new
partNumber.  As expected the {_.version) value changed to "1" for the new
design, but the {_.file[-3:]} value remained the same as before.  I can
live with that, as a work-a-round, it can be fixed by;
temporarily modifying parameter line to something like "Tmp{_.file[-3:]}"
and then back as desired.  But it would be nice if "file" value changed on
"Save As"  and the work-a-around was not necessary.

Thanks for your efforts.

Regards,
Eldon Brown

**Comment by thomasa88** [comment]
I remember trying to get the correct value for `_.file` on a rename, but I failed. As far as I recall, the problem is that the filename is set at (or after) the save event, meaning that the add-in has no way of knowing the new name until after the file has been saved.

I opted to not update the value after the save, as updating the value would lead to the file being "modified", even though the user just saved. (There is also the interesting case of the program exiting/crashing just after saving the file, or the user not saving the modified file. Should the add-in correct the name every time the file is opened? (and make it "modified"). Possibly, the add-in could try to force a save after such a modification.)

This is the current workaround, from the README:
> _.file does not update on file rename/copy.
> 
>    Workaround is to open the Change Text Parameters and click OK in the new document.

I might rethink this decision in the future or if there are many requests for a changed behavior.
